### PR TITLE
Implement zerolog as logger, support proper json log format and human…

### PR DESCRIPTION
**Pull request by MarianZoll**
** Describe the change **

Implement a new logger that support json and text logging as well as human-readable log levels out of the box. Additionally, the logger can be extended based on new features of https://github.com/rs/zerolog without all the overhead and manual effort required for glog which is used so far.

** Issue reference **

https://github.com/kiali/kiali/issues/2893 

And pull requests for the helm-charts and operator repo:
https://github.com/kiali/helm-charts/pull/15
https://github.com/kiali/kiali-operator/pull/187

** Backwards incompatible? **

No, based on the discussions we had, we made sure that the `deployment.verbose_mode` setting is still supported. Nothing else should be affected.

** Documentation **

The documentation is added in the operator CR:

https://github.com/kiali/kiali-operator/pull/187/files#diff-b05613685dff7d8a5eb70946010acb19192f6f9edc83d292af21db46aa9be89dR280-R286

In case additional documentation is required somewhere, please let me know.

* Does your contribution contain user-visible changes like e.g. new or changed configuration settings?

Should not. Only the log output will look different, but all information should stay the same.

### CODE REVIEW

File: log/log.go
Comment (jmazzitelli) [inline log/log.go:6]: trivial - the imports in the kiali src code base are formatted specifically - base libs first, blank line, third party libs, blank line, kiali libs last. So this should be:

```suggestion
	"os"

	"github.com/rs/zerolog"
	"github.com/rs/zerolog/log"
```

File: log/log.go
Comment (jmazzitelli) [inline log/log.go:43]: we can remove this comment as numeric log levels are not used anymore
```suggestion
```

File: log/log.go
Comment (jmazzitelli) [inline log/log.go:48]: we can remove this comment as numeric log levels are not used anymore
```suggestion
```

File: log/log.go
Comment (jmazzitelli) [inline log/log.go:57]: we can remove this comment as numeric log levels are not used anymore
```suggestion
```

File: log/log.go
Comment (jmazzitelli) [inline log/log.go:62]: we can remove this comment as numeric log levels are not used anymore
```suggestion
```

File: log/log.go
Comment (jmazzitelli) [inline log/log.go:99]: I think these levels are incorrect. Based on the current [Kiali CR documentation](https://github.com/kiali/kiali-operator/blob/v1.24/deploy/kiali/kiali_cr.yaml#L355-L358), where it says,

```
Determines which priority levels of log messages Kiali will output.
# Typical values are "3" for INFO and higher priority, "4" for DEBUG and higher priority.
```
This is based on the glog "V" levels we used before, higher numbers get noisier as you go up. So in Kiali, 3 is INFO, 4 is DEBUG and 5 is TRACE , with 2 being WARN, and 1 being ERROR and 0 being FATAL.
So Kiali users will be using lower numbers to make the log quieter.

```suggestion
	case "0":
		return zerolog.FatalLevel
	case "1":
		return zerolog.ErrorLevel
	case "2":
		return zerolog.WarnLevel
	case "3":
		return zerolog.InfoLevel
	case "4":
		return zerolog.DebugLevel
	case "5":
		return zerolog.TraceLevel
```

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
Some trivial things, except for the verbose_mode conversion. That looks wrong. See below.

**Review by jmazzitelli — COMMENTED** [review]
Looks like a conflict in go.mod - probably a minor thing that needs to be fixed.

**Comment by MarianZoll** [comment]
Sorry missed that one this morning.

**Comment by MarianZoll** [comment]
But seems like Travis as some issues with linting. I will make sure that the build passes.

**Comment by MarianZoll** [comment]
Linting should be fine now. Not sure why `make clean build test-race` fails due to permission errors. But locally it seems to work fine.

**Comment by jmazzitelli** [comment]
(I'll try to investigate the travis failures - I doubt it has anything to do with this PR).

**Comment by jmazzitelli** [comment]
travis failure has been corrected. I'll be testing this stuff soon

**Review by jmazzitelli — COMMENTED** [review]
I'm running the molecule tests on these now. I see the logs now look like this:

```
7:28PM INF Kiali: Version: v1.27.0-SNAPSHOT, Commit: 1f3247892d60945421d92cec16a63f499a85005c

7:28PM INF Using authentication strategy [anonymous]
7:28PM WRN Kiali auth strategy is configured for anonymous access - users will not be authenticated.
```
Notice the timestamps are less-than-helpful because they are only at granularity of minutes (`7:28PM` rather than, say, `15:28:15.032` - i.e. 24-hour clock at a more fine-grained milliseconds).

Having at minimum seconds in the timestamp (if not millis) allows one to easily refer to a line in the logs ("Hey, go check out the logs at 7:28:15") and you can roughly see how fast things are going (on the order of at least seconds).

Is there some way we can configure the timestamps (either hardcode it in the log.go module or maybe even add another `logger` setting?)

**Comment by MarianZoll** [comment]
Sure, I will make some changes regarding the timestamp and probable other features that might be nice to include. I will attach a snippet to this PR and will update the operator and helm charts accordingly.

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
I would also like a simple unit test `log/log_test.go` here. I just wrote up the start of one that you should use to start with. Note that I am getting failures when I set the LOG_LEVEL env. var. to non-numeric level names (tests pass when using the numerics). I haven't looked at the failures yet - I'm just writing a simple unit test that I thought should just work.

[log_test.go.txt](https://github.com/kiali/kiali/files/5477882/log_test.go.txt)

I'm getting errors like:

```
    log_test.go:73: 
        	Error Trace:	log_test.go:73
        	Error:      	Not equal: 
        	            	expected: 2
        	            	actual  : 1
        	Test:       	TestEnvVarLogLevel
        	Messages:   	LOG_LEVEL=WARN,levelToTest={stringLevel:WARN level:2}
```

Which means I'm setting the env var LOG_LEVEL to "WARN" but the resolveLogLevelFromEnv() is always returning LOG_LEVEL=INFO. And I'm seeing a warning log message spit out on these failures saying `{"level":"warn","time":"2020-11-02T15:33:47-05:00","message":"Provided LOG_LEVEL WARN is invalid. Fallback to info."}`

File: log/log.go
Comment (jmazzitelli) [inline log/log.go:4]: Add "strings" to get the `ToLower` function
```suggestion
	"os"
	"strings"
```

File: log/log.go
Comment (jmazzitelli) [inline log/log.go:98]: Convert to lowercase since that is what the parser wants. Some people are going to think you need to specify the level as all uppercase (e.g. "WARN" or "DEBUG") but this wants all lower. Just support any case by converting to all lower:
```suggestion
		logLevelFromString, err := zerolog.ParseLevel(strings.ToLower(logLevel))
```

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
Convert to all lowercase when needing to parse the log level name.

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
Also, as mentioned [here](https://github.com/kiali/kiali-operator/pull/187#issuecomment-720732741), there are two places we have to clean up. Delete the setting and usage of VERBOSE_MODE env var in the makefiles. We don't need to set that via env var - devs can manually set it if they need to. So these need to be deleted:

- [ ] Delete line from make/Makefile.operator.mk: https://github.com/kiali/kiali/blob/master/make/Makefile.operator.mk#L72
- [ ] Delete line from Makefile: https://github.com/kiali/kiali/blob/master/Makefile#L113

**Comment by MarianZoll** [comment]
Short summary of the latest commits:

- Implemented the changes you supposed. 
- Added the possibility for log sampling based on another environment variable. That might be a reasonable feature for productive system that want to reduce the log volume collected by external ops tools. By default, this feature will be disabled. 
- Extended the `log_test.go` you provided (Thank you!) with some regression tests for the logger itself to ensure that changes to the logging process would be identified as early as possible. 

In case you agree with the made changes, I will incorporate the settings for `LOG_TIME_FIELD_FORMAT` and `LOG_SAMPLER_RATE` to the helm chart and operator and update the PRs accordingly.

File: log/log.go
Comment (jmazzitelli) [inline log/log.go:19]: Is there some way to parse the time field format for correctness? Similar to what you do for LOG_SAMPLER_RATE down below when you confirm it can be parsed as a number, and what you do for LOG_LEVEL in resolveLogLevelFromEnv().

Perhaps zerolog has a way to check it for validity. I'm thinking what happens if a person enters some invalid time field - we would want this to either fail-fast or default to some valid format. Something like:
```suggestion
	} else if !zerolog.IsTimeFieldFormatValid(logTimeFieldFormat) { // I have no idea if that API exists - but you get the idea
		log.Warn().Msgf("Provided time field format [%s] is not valid. Defaulting to RFC3339.", logTimeFieldFormat)
		logTimeFieldFormat = time.RFC3339
	}
```

File: log/log.go
Comment (jmazzitelli) [inline log/log.go:36]: Again, I wonder if we should test for validity of the LOG_FORMAT using some zerolog parse function?

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
> In case you agree with the made changes, I will incorporate the settings for `LOG_TIME_FIELD_FORMAT` and `LOG_SAMPLER_RATE` to the helm chart and operator and update the PRs accordingly.

Yes, I like this. Let's incorporate the new settings in the helm chart and operator PRs accordingly. Aside from my comments here in this review, this all looks good. 

If you can incorporate those validity checks in the code which I talk about in this review, that's great; but if there is no easy way to perform those validity checks, don't worry about it. In other words, my requested changes here are minor and low priority. You tell me if it is easy to incorporate those changes - if not, I'll remove my "request for changes" in this PR and we can move forward with the review/approval process to get this merged soon.

Thanks for all the help!

File: log/log_test.go
Comment (jmazzitelli) [inline log/log_test.go:135]: Probably should be careful here. You are clearing the entire environment which is going to affect all future unit tests that are run after this one executes. You should consider using a "defer" here and rebuild the environment back to the way it was before this test function executes so future unit tests don't lose their environment. So far, all unit tests pass, but this could get ugly in the future with this bug being VERY hard to find ("where did all my env vars go??") :)

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
Consider fixing this as well.

File: log/log_test.go
Comment (jmazzitelli) [inline log/log_test.go:131]: Because the test is clearing the entire environment, it will affect all future unit tests that are run after this one executes. So far, all unit tests pass, but this could get ugly in the future with this bug being VERY hard to find ("where did all my env vars go??") :)

Here's a fix - using "defer", this unit test will put the environment back to the way it was before this test function executes so future unit tests don't lose their environment. 

```suggestion

	originalEnv := os.Environ()
	defer func() {
		for _, kvp := range originalEnv {
			kvpArray := strings.Split(kvp, "=")
			os.Setenv(kvpArray[0], kvpArray[1])
		}
		return
	}()
```

**Review by jmazzitelli — CHANGES_REQUESTED** [review]

File: log/log.go
Comment (MarianZoll) [reply log/log.go:19]: Hm it does rely on the standard time package (https://golang.org/pkg/time/) which to the best of my knowledge could take any malformed time format and parse it to a non-sense time object. 

I tried to do a couple of format checks on the current timestamp ála:
```
timestamp := time.Now()
parsedAsString := timestamp.Format(format)
parsedAsTimestamp, err := time.Parse(format, parsedAsString)
```
but this seems to be unlikely successful as the format can range from `stupidInput`, `2006` to a standard format like `2006-01-02T15:04:05Z07:00`. Especially the fact that you can provide wrong formats like `2020` which will completely mess up the logic (parses the year to 9090) will probably brake the check routine. 

Neither the time module itself nor zerolog has validators for the format. What we might want to consider is that we only support a subset of time formats like the ones shipped with the time module:
```
	ANSIC       = "Mon Jan _2 15:04:05 2006"
	UnixDate    = "Mon Jan _2 15:04:05 MST 2006"
	RubyDate    = "Mon Jan 02 15:04:05 -0700 2006"
	RFC822      = "02 Jan 06 15:04 MST"
	RFC822Z     = "02 Jan 06 15:04 -0700" // RFC822 with numeric zone
	RFC850      = "Monday, 02-Jan-06 15:04:05 MST"
	RFC1123     = "Mon, 02 Jan 2006 15:04:05 MST"
	RFC1123Z    = "Mon, 02 Jan 2006 15:04:05 -0700" // RFC1123 with numeric zone
	RFC3339     = "2006-01-02T15:04:05Z07:00"
	RFC3339Nano = "2006-01-02T15:04:05.999999999Z07:00"
	Kitchen     = "3:04PM"
	// Handy time stamps.
	Stamp      = "Jan _2 15:04:05"
	StampMilli = "Jan _2 15:04:05.000"
	StampMicro = "Jan _2 15:04:05.000000"
	StampNano  = "Jan _2 15:04:05.000000000"
```

We could check against those and fallback to `RFC3339` in case the validations fails. What do you think?

File: log/log.go
Comment (MarianZoll) [reply log/log.go:36]: Guess this is already some kind of validation. Everything else than `json` will result in a regular text output. We could validate against an enum of text and json to give some debugging info to the user. Guess that might be reasonable to do.

**Comment by MarianZoll** [comment]
I fixed the environment setup and validated the `LOG_FORMAT`. I adopt the `LOG_TIME_FIELD_FORMAT` based on our discussion later on.

File: log/log.go
Comment (MarianZoll) [reply log/log.go]: Proposal see below.

**Review by jmazzitelli — COMMENTED** [review]
OK, I ran this and it all works. Thanks @MarianZoll !

I have one thing I want to think about before giving this the final approval. It is probably a minor thing - but I want to think about it. If anyone has suggestions or comments, feel free to do so.

It has to do with the colored fonts emitted by zerolog. If you look at the logs within the k8s console (in my case, the OpenShift Console), the log messages are ugly and hard to read due to the color ASCII codes embedded in the log messages. For example:
image

While that is ugly, the `kubectl logs` (or `oc logs`) command will output the color-codes, too, but if your command line window can support them, the output looks nice.

image

So I'm not sure what we should do or which log output we should cater to (console window or command line window?).

Is it possible to add yet another config option to let the user choose to emit colored logs or not? I'm not sure if zerolog has that config option, or, even if they do, should we care about this?

**Comment by jmazzitelli** [comment]
Interestingly, if you change to "json" format, it actually looks better in the console:
image
so maybe that is the "solution". If someone doesn't like the color-codes (in case they are not interpreted correctly by the viewer and you see the codes rather than the colors), they can switch to 'json' and it will look "better" albeit with a little more noise (the "keys" in the json)

**Comment by jmazzitelli** [comment]
FWIW: I'm leaning towards leaving everything as-is (rather than bothering to add more config options). Users can switch to 'json' format if they'd like.

So I am inclined to merge this as-is, but I'll wait to hear other's opinions before doing so.

**Comment by jshaughn** [comment]
@jmazzitelli I don't think we want colorization at all.  It could generate unexpected results in any non-terminal log display (e.g. OS console, Kiali).  In Kiali it could also confuse parsing, or regex-based show/hide. 

I don't think it needs an option, I think it should be turned off.

**Comment by jmazzitelli** [comment]
> I don't think it needs an option, I think it should be turned off.

OK, it looks easy to do that:
```
log := zerolog.New(zerolog.ConsoleWriter{Out: os.Stdout, NoColor: true})
```
I'll add a suggested fix so @MarianZoll can change it quickly. Should be a simple one-line change.

File: log/log.go
Comment (jmazzitelli) [inline log/log.go:56]:
```suggestion
		log.Logger = log.Output(zerolog.ConsoleWriter{Out: os.Stdout, TimeFormat: zerolog.TimeFieldFormat, NoColor: true})
```

**Review by jmazzitelli — CHANGES_REQUESTED** [review]
We'll want to turn off colorization for the "text" format.

File: log/log.go
Comment (jmazzitelli) [reply log/log.go:56]: This is the final change. Everything else is good and working. Once this change is incorporated, we can merge the three PRs.

**Comment by MarianZoll** [comment]
Wanted to propose the same. Simply turning it off sounds reasonable to do. I committed your suggestion. 

@jmazzitelli Thank you for guiding and reviewing the PR. This was really fun implementing and we are looking forward using the features in multiple projects to optimize the day-2 ops of the apps and opps stack! 

Let me know in case something else is needed..

**Review by jmazzitelli — APPROVED** [review]

**Pull request merged**
